File: servers/typescript/src/core/canonicalize.ts
// Output canonicalizer.
//
// Two responses are "equivalent" if their canonical forms are byte-equal.
// Used by the Phase-0.5 Py↔Py parity harness and, in later phases, by
// the TS↔Py byte-diff CI gate.
//
// Normalizations applied:
//   - Strip transient keys (latencies, timestamps, transcript paths).
//   - Replace ISO-8601 timestamps with the fixed token `<TS>`.
//   - Replace UUIDs with positional tokens `<UUID:N>` in first-seen order.
//   - Replace canary nonces (`CC_CANARY_<HEX>`) with `<CANARY:N>`.
//   - Replace session ids matching `s-XXXX` with `<SID:N>`.
//   - Round numbers to 6 decimal places (string `.toFixed(6)` semantics).
//   - Sort object keys recursively (Unicode lexicographic).
//
// All of this must match the Python mirror in `scripts/canonicalize.py`
// EXACTLY. Each change here needs a matching change there.

const TRANSIENT_KEYS = new Set<string>([
  // Timing — never deterministic.
  "wall_ms",
  "cpu_ms",
  "elapsed_ms",
  "wall_used_ms",
  "wall_remaining_ms",
  "cpu_used_ms",
  // Timestamps embedded in run_summary / sessions / pin files.
  "started_at",
  "ended_at",
  "pinned_at",
  "last_at",
  "created_at",
  "stale_at",
  "ts",
  // Transcript paths embed millisecond filenames.
  "transcript_path",
  "transcript",
  // Cache-hit counters — true cross-run but accumulate; tests rarely care.
  "cache_hits",
  // Per-call HTTP attempts vary with transient network.
  "attempts",
]);

const ISO_TS_RE   = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?Z?$/;
const UUID_RE     = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
const CANARY_RE   = /^CC_CANARY_[0-9A-F]+$/;
const SID_RE      = /^s-[0-9a-f]{4,}$/;

const ISO_INLINE_RE    = /\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?Z?/g;
const UUID_INLINE_RE   = /[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/gi;
const CANARY_INLINE_RE = /CC_CANARY_[0-9A-F]+/g;
const SID_INLINE_RE    = /\bs-[0-9a-f]{4,}\b/g;

export interface CanonicalizeOptions {
  /** Number of decimal places to round floats to before serializing.
   *  Default 6 — matches Python's mirror. */
  floatPrecision?: number;
  /** Override transient-key set. Useful when a specific test wants
   *  timing assertions and explicitly opts back in. */
  transientKeys?: ReadonlySet<string>;
}

/** Canonicalize a JS value and return a stable JSON string. The string
 *  is suitable for direct byte-comparison. */
export function canonicalize(value: unknown, opts?: CanonicalizeOptions): string {
  const ctx: Ctx = {
    floatPrecision: opts?.floatPrecision ?? 6,
    transient: opts?.transientKeys ?? TRANSIENT_KEYS,
    uuids: new Map(),
    canaries: new Map(),
    sids: new Map(),
  };
  return jsonify(normalize(value, ctx));
}

interface Ctx {
  floatPrecision: number;
  transient: ReadonlySet<string>;
  uuids: Map<string, string>;
  canaries: Map<string, string>;
  sids: Map<string, string>;
}

function normalize(v: unknown, ctx: Ctx): unknown {
  if (v === null || v === undefined) return null;
  if (Array.isArray(v)) return v.map((x) => normalize(x, ctx));
  if (typeof v === "number") {
    if (!Number.isFinite(v)) return null;
    if (Number.isInteger(v)) return v;
    // Use toFixed(N) so 1.1+2.2 doesn't render as 3.3000000000000003.
    return Number(v.toFixed(ctx.floatPrecision));
  }
  if (typeof v === "string") return normalizeString(v, ctx);
  if (typeof v === "boolean") return v;
  if (typeof v === "object") {
    const out: Record<string, unknown> = {};
    const entries = Object.entries(v as Record<string, unknown>);
    for (const [k, val] of entries) {
      if (ctx.transient.has(k)) continue;
      out[k] = normalize(val, ctx);
    }
    return out;
  }
  // Functions / symbols / bigint shouldn't appear in MCP outputs; if they
  // do, stringify defensively.
  return String(v);
}

function normalizeString(s: string, ctx: Ctx): string {
  // Exact-match token replacements first (faster + clearer).
  if (ISO_TS_RE.test(s))   return "<TS>";
  if (UUID_RE.test(s))     return tokenFor(ctx.uuids, s, "UUID");
  if (CANARY_RE.test(s))   return tokenFor(ctx.canaries, s, "CANARY");
  if (SID_RE.test(s))      return tokenFor(ctx.sids, s, "SID");
  // Inline replacements — strings that EMBED a UUID / TS / canary.
  let out = s;
  out = out.replace(ISO_INLINE_RE, "<TS>");
  out = out.replace(UUID_INLINE_RE, (m) => tokenFor(ctx.uuids, m, "UUID"));
  out = out.replace(CANARY_INLINE_RE, (m) => tokenFor(ctx.canaries, m, "CANARY"));
  out = out.replace(SID_INLINE_RE, (m) => tokenFor(ctx.sids, m, "SID"));
  return out;
}

function tokenFor(map: Map<string, string>, key: string, prefix: string): string {
  const existing = map.get(key);
  if (existing) return existing;
  const tok = `<${prefix}:${map.size + 1}>`;
  map.set(key, tok);
  return tok;
}

/** JSON-serialize with sorted keys. The replacer is run AFTER normalize()
 *  has already pruned transient keys, so the only job here is key order. */
function jsonify(v: unknown): string {
  return JSON.stringify(v, sortReplacer, 0);
}

function sortReplacer(_k: string, val: unknown): unknown {
  if (val !== null && typeof val === "object" && !Array.isArray(val)) {
    const sorted: Record<string, unknown> = {};
    for (const k of Object.keys(val as Record<string, unknown>).sort()) {
      sorted[k] = (val as Record<string, unknown>)[k];
    }
    return sorted;
  }
  return val;
}

File: scripts/canonicalize.py
#!/usr/bin/env python3
"""Output canonicalizer — Python mirror.

MUST stay byte-identical with `servers/typescript/src/core/canonicalize.ts`.
Every normalization decision (transient-key set, ISO/UUID/canary/SID
patterns, float precision, key-sort order) is intentionally duplicated
so the two halves of the Py↔Py / TS↔Py parity gate produce the SAME
canonical bytes given the SAME logical state.

When you change one, change the other in lockstep.
"""

from __future__ import annotations

import json
import re
import sys
from typing import Any, Mapping

# These match TRANSIENT_KEYS in the TS canonicalizer exactly.
_TRANSIENT_KEYS: frozenset[str] = frozenset({
    # Timing
    "wall_ms", "cpu_ms", "elapsed_ms",
    "wall_used_ms", "wall_remaining_ms", "cpu_used_ms",
    # Timestamps
    "started_at", "ended_at", "pinned_at", "last_at", "created_at",
    "stale_at", "ts",
    # Transcript paths
    "transcript_path", "transcript",
    # Cache-hit counters
    "cache_hits",
    # HTTP retry counts
    "attempts",
})

_ISO_TS_RE    = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?Z?$")
_UUID_RE      = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$",
    re.IGNORECASE,
)
_CANARY_RE    = re.compile(r"^CC_CANARY_[0-9A-F]+$")
_SID_RE       = re.compile(r"^s-[0-9a-f]{4,}$")

_ISO_INLINE_RE    = re.compile(r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?Z?")
_UUID_INLINE_RE   = re.compile(
    r"[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}",
    re.IGNORECASE,
)
_CANARY_INLINE_RE = re.compile(r"CC_CANARY_[0-9A-F]+")
_SID_INLINE_RE    = re.compile(r"\bs-[0-9a-f]{4,}\b")


def canonicalize(value: Any, *, float_precision: int = 6,
                  transient: frozenset[str] | None = None) -> str:
    """Canonicalize a JSON-like value to a stable string suitable for byte
    comparison against TypeScript's canonicalize() output."""
    ctx = _Ctx(
        float_precision=float_precision,
        transient=transient if transient is not None else _TRANSIENT_KEYS,
        uuids={},
        canaries={},
        sids={},
    )
    normalized = _normalize(value, ctx)
    # sort_keys + separators=(",", ":") matches JSON.stringify(v, replacer, 0)
    # with our sortReplacer. ensure_ascii=False so embedded unicode survives.
    return json.dumps(normalized, sort_keys=True, separators=(",", ":"),
                       ensure_ascii=False)


class _Ctx:
    __slots__ = ("float_precision", "transient", "uuids", "canaries", "sids")

    def __init__(self, float_precision: int, transient: frozenset[str],
                  uuids: dict[str, str], canaries: dict[str, str],
                  sids: dict[str, str]):
        self.float_precision = float_precision
        self.transient       = transient
        self.uuids           = uuids
        self.canaries        = canaries
        self.sids            = sids


def _normalize(v: Any, ctx: _Ctx) -> Any:
    if v is None:
        return None
    if isinstance(v, bool):           # must check BEFORE int (bool is int)
        return v
    if isinstance(v, int):
        return v
    if isinstance(v, float):
        # Match JS Number.isFinite() / toFixed(N) semantics. NaN / Inf -> null.
        if v != v or v in (float("inf"), float("-inf")):
            return None
        if v.is_integer():
            return int(v)
        # Round to N decimal places, then re-parse so trailing zeros are
        # dropped (matching JS `Number(x.toFixed(6))`).
        return float(f"{v:.{ctx.float_precision}f}")
    if isinstance(v, str):
        return _normalize_str(v, ctx)
    if isinstance(v, list) or isinstance(v, tuple):
        return [_normalize(x, ctx) for x in v]
    if isinstance(v, Mapping):
        out: dict[str, Any] = {}
        for k, val in v.items():
            if k in ctx.transient:
                continue
            out[str(k)] = _normalize(val, ctx)
        return out
    # bytes / set / etc. — stringify defensively, matching TS String(v).
    return str(v)


def _normalize_str(s: str, ctx: _Ctx) -> str:
    if _ISO_TS_RE.match(s):
        return "<TS>"
    if _UUID_RE.match(s):
        return _token_for(ctx.uuids, s, "UUID")
    if _CANARY_RE.match(s):
        return _token_for(ctx.canaries, s, "CANARY")
    if _SID_RE.match(s):
        return _token_for(ctx.sids, s, "SID")
    out = s
    out = _ISO_INLINE_RE.sub("<TS>", out)
    out = _UUID_INLINE_RE.sub(lambda m: _token_for(ctx.uuids, m.group(0), "UUID"), out)
    out = _CANARY_INLINE_RE.sub(lambda m: _token_for(ctx.canaries, m.group(0), "CANARY"), out)
    out = _SID_INLINE_RE.sub(lambda m: _token_for(ctx.sids, m.group(0), "SID"), out)
    return out


def _token_for(table: dict[str, str], key: str, prefix: str) -> str:
    existing = table.get(key)
    if existing is not None:
        return existing
    tok = f"<{prefix}:{len(table) + 1}>"
    table[key] = tok
    return tok


# ----------------------------------------------------------------------
# CLI: `canonicalize.py <file.jsonl>` prints each line's canonical form.
# Useful for debugging diff failures.
# ----------------------------------------------------------------------
def _cli(argv: list[str]) -> int:
    if len(argv) != 2:
        sys.stderr.write("usage: canonicalize.py <file.jsonl>\n")
        return 2
    with open(argv[1], "r", encoding="utf-8") as f:
        for line in f:
            line = line.rstrip("\n")
            if not line:
                print("")
                continue
            try:
                doc = json.loads(line)
            except json.JSONDecodeError:
                # Mirror TS safeParse: compare raw text on parse failure.
                doc = line
            print(canonicalize(doc))
    return 0


if __name__ == "__main__":
    sys.exit(_cli(sys.argv))
